client/rpc/rpc_test.go

```diff
@@ -60,40 +61,47 @@ func (s *IntegrationTestSuite) TestCLIQueryConn() {
 
 func (s *IntegrationTestSuite) TestQueryABCIHeight() {
 	testCases := []struct {
-		name      string
-		reqHeight int64
-		ctxHeight int64
-		expHeight int64
+		name                string
+		reqHeight           int64
+		ctxHeight           int64
+		awaitMinChainHeight int64
+		assertFn            func(t *testing.T, latestHeightAtQuery int64, resp abci.ResponseQuery)
 	}{
 		{
-			name:      "non zero request height",
-			reqHeight: 3,
-			ctxHeight: 1, // query at height 1 or 2 would cause an error
-			expHeight: 3,
+			name:                "request height set",
+			reqHeight:           2, // no proof when < 2
+			ctxHeight:           1,
+			awaitMinChainHeight: 3, // wait +1 block to be on the safe side
+			assertFn: func(t *testing.T, _ int64, resp abci.ResponseQuery) {
+				assert.Equal(t, int64(2), resp.Height)
+			},
 		},
 		{
-			name:      "empty request height - use context height",
-			reqHeight: 0,
-			ctxHeight: 3,
-			expHeight: 3,
+			name:                "fallback to context height when request height is not set",
+			reqHeight:           0,
+			ctxHeight:           3,
+			awaitMinChainHeight: 4, // wait +1 block to be on the safe side
+			assertFn: func(t *testing.T, _ int64, resp abci.ResponseQuery) {
+				assert.Equal(t, int64(3), resp.Height)
+			},
 		},
 		{
-			name:      "empty request height and context height - use latest height",
-			reqHeight: 0,
-			ctxHeight: 0,
-			expHeight: 4,
+			name:                "with empty values, use latest height",
+			reqHeight:           0,
+			ctxHeight:           0,
+			awaitMinChainHeight: 2, // no proof when < 2
+			assertFn: func(t *testing.T, latestHeightAtQuery int64, resp abci.ResponseQuery) {
+				anyOf := []int64{latestHeightAtQuery, latestHeightAtQuery - 1}
```

Comment:
Sometimes the server returned `latestHeightAtQuery - 1` in the reponse. This was very unexpected. 
`client ->GetLatestBlock` seems to be updated while the internal state in [baseapp](https://github.com/cosmos/cosmos-sdk/blob/release/v0.53.x/baseapp/abci.go#L163) was not.

I have modified the accept criteria for the test as IMHO the main focus is on the default for the missing height field in the request.